### ISSUE TITLE
struct.pack / struct.unpack / struct.calcsize on fields with variable size leaks memory in Python 2

### ISSUE THREAD

**Issue by mgrrx** [ros/genpy#128]
Originally caused by https://bugs.python.org/issue14596

TL;DR: The python struct module caches struct instances.

All calls of the module functions struct.pack, struct.unpack and struct.calcsize make use of this cache.
https://github.com/ros/genpy/blob/kinetic-devel/src/genpy/generator.py

We debugged this issue with a simple script like this and a huge map published on /map:

```
import rospy
import nav_msgs.msg
rospy.init_node("test")
def noop(msg):
    print "msg arrived"
s = rospy.Subscriber("/map", nav_msgs.msg.OccupancyGrid, noop)
rospy.sleep(5)
s.unregister()
rospy.spin()
```

The result is that the node still consumes 370 MB after the print of "msg arrived" and even after unregistering the subscriber (to rule out latched topic msg caching).

Calling `struct._clearcache()` frees the RAM and we get down to 52 MB.

### Solution
Calls of struct.pack, struct.unpack and struct.callback need to be substituted by the respective calls of the Struct class.
